codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -121,35 +123,84 @@ public void close() { }
     @Override
     public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf data,
             int padding, boolean endStream, ChannelPromise promise) {
-        boolean releaseData = true;
+        boolean lastHeaderWritten = false;
+        boolean lastFrameWritten = false;
         SimpleChannelPromiseAggregator promiseAggregator =
                 new SimpleChannelPromiseAggregator(promise, ctx.channel(), ctx.executor());
+
+        // All padding will be put at the end, so in the worst case we need 3 headers:
+        // a repeated no-padding frame of maxFrameSize, a frame that has part data and part
+        // padding, and a frame that has the remainder of the padding.
+        ByteBuf headerBuf = ctx.alloc().buffer(3 * DATA_FRAME_HEADER_LENGTH);
         try {
             verifyStreamId(streamId, STREAM_ID);
             verifyPadding(padding);
 
-            Http2Flags flags = new Http2Flags().paddingPresent(padding > 0).endOfStream(endStream);
+            int remainingData = data.readableBytes();
+            int headerIndex = 0;
+            int prevFrameDataBytes = 0;
+            int prevFramePaddingBytes = 0;
+            ByteBuf frameHeader = null;
 
-            int payloadLength = data.readableBytes() + padding + flags.getPaddingPresenceFieldLength();
-            verifyPayloadLength(payloadLength);
+            do {
+                // Determine how much data and padding to write in this frame.
+                int frameDataBytes = min(remainingData, maxFrameSize);
+                int framePaddingBytes = min(padding, max(0, (maxFrameSize - 1) - frameDataBytes));
+
+                // Decrement the remaining counters.
+                padding -= framePaddingBytes;
+                remainingData -= frameDataBytes;
+                boolean lastFrame = remainingData == 0 && padding == 0;
+
+                if (frameDataBytes != prevFrameDataBytes || framePaddingBytes != prevFramePaddingBytes
+                        || frameHeader == null) {
+                    // Go to the next unused header frame and write it.
+                    frameHeader = headerBuf.slice(headerIndex++ * DATA_FRAME_HEADER_LENGTH,
+                            DATA_FRAME_HEADER_LENGTH).writerIndex(0);
+                    prevFrameDataBytes = frameDataBytes;
+                    prevFramePaddingBytes = framePaddingBytes;
+                    Http2Flags flags = new Http2Flags().paddingPresent(framePaddingBytes > 0);
+                    if (endStream && lastFrame) {
+                        // This is the last frame to be sent for this stream.
+                        flags.endOfStream(true);
+                    }
+
+                    int payloadLength = frameDataBytes + framePaddingBytes + flags.getPaddingPresenceFieldLength();
+                    verifyPayloadLength(payloadLength);
```

Comment:
Is this call necessary now that we're enforcing the frame length restriction above anyway?